### PR TITLE
efi-api.c: check availability of TPM2 device also from kernel TPM eve…

### PR DESCRIPTION
**Pull request by mikkorapeli-linaro**
…nt log

If EFI firmware has TPM support, they very often export the TPM EventLog to Linux kernel. Kernel can export this information to sysfs and userspace can check for it and if found, queue tpm2.target for loading TPM kernel modules etc also in early boot. Note that TPM device support for EventLog is optional but very useful when investigating PCR signatures and their changes so most HW provide this.

Kernel patch to expose TPM event log via sysfs has been proposed here:
https://lore.kernel.org/all/20240422112711.362779-1-mikko.rapeli@linaro.org/T/#u

Fixes detection of TPM2 device on qemu with swtpm when TPM kernel drivers are compiled as modules and loaded by udev. Support is correctly detected and tpm2.target triggered. Without this patch TPM2 device support needs to be visible from the firmware ACPI tables or the kernel drivers need to be compiled into the kernel.

See also:
https://lists.freedesktop.org/archives/systemd-devel/2024-April/050206.html
https://github.com/systemd/systemd/pull/32314

### CODE REVIEW

File: src/shared/efi-api.c [hunk 1]
```diff
@@ -503,8 +503,19 @@ bool efi_has_tpm2(void) {
         if (errno != ENOENT)
                 log_debug_errno(errno, "Unable to test whether /sys/firmware/acpi/tables/TPM2 exists, assuming it doesn't: %m");
 
-        /* As the last try, check if the EFI firmware provides the EFI_TCG2_FINAL_EVENTS_TABLE
-         * stored in EFI configuration table, see:
+        /* If EFI firmware used TPM to e.g. measure loaded binaries, it can export TPM EventLog to
+         * Linux kernel, which is a strong indication that TPM device is available and used.
+         * Then a patched Linux kernel can expose this information to userspace so that we know
+         * to queue TPM device startup and module loading. */
+        cache = access("/sys/firmware/efi/tpm_log", F_OK) >= 0;
```
Comment (poettering) [inline src/shared/efi-api.c:510]: so this would show up on all systems with efi firmware? that'd be nice.

Comment (mikkorapeli-linaro) [reply src/shared/efi-api.c:510]: On all EFI devices where TPM device is used by firmware and the TPM Event Log is generated. This seems to be most devices. Event log is optional in TPM standards but it makes any troubleshooting with TPM a lot easier so most seem to support this.

**Comment by mikkorapeli-linaro** [comment]
Yes, kernes patch has just been posted so this should wait for it. Kernel do export other similar variables in /sys/firmware/efi/systab but that interface should not be extended according comments in code so I proposed a new sysfs file for this notification. This works on qemu+swtpm, rockpi4b (fTPM), synquacer (fTPM) etc arm64 boards which don't provide ACPI tables but only devicetree entries for TPM devices.

**Comment by poettering** [comment]
ok, looks good to me codewise. please report if this entered the right git tree that makes it clear that this is very likely going to enter Linus' tree soon. That's enough for us, and we can merge this here, too.

Please update the commit msg with a link to the merged upstream commit then.

**Comment by mikkorapeli-linaro** [comment]
Yes, I will update the kernel patch status here.

**Comment by poettering** [comment]
and then also mention in a comment next to the check which kernel version this test requires at minimum.

**Comment by poettering** [comment]
@mikkorapeli-linaro did this ever make it into the upstream kernel?

**Comment by poettering** [comment]
Ah, judging from reading that thread again I sense this is not going to land, and at least on ACPI we should be good anyway, because we can check for existance of that table via sysfs directly.

Hence closing, I guess? if things change on this, please let us know, we can reopen this.

**Comment by mikkorapeli-linaro** [comment]
Yes, I can re-open if/when kernel patch gets merged. A solution for Arm System Ready IR (embedded) devices without ACPI is needed.

**Pull request closed without merging**